File: test/corpus/propagation/fallible_option_case.exs
defmodule RustQ.Corpus.Propagation.FallibleOptionCase do
  @moduledoc "Fallible option scrutinees propagate before option-case lowering."

  use RustQ.Meta

  alias RustQ.Type, as: R

  @spec maybe_size(atom()) :: R.nif_result(R.option(R.f32()))
  defrust maybe_size(_key) do
    {:ok, {:some, 12.0}}
  end

  @spec apply_size(atom()) :: R.nif_result(R.f32())
  defrust apply_size(key) do
    value = 1.0

    case maybe_size(key) do
      {:some, size} -> value = size
      :none -> :ok
    end

    {:ok, value}
  end

  @spec default_size(atom()) :: R.nif_result(R.f32())
  defrust default_size(key) do
    {:ok, maybe_size(key).unwrap_or(1.0)}
  end
end

File: test/corpus/propagation/fallible_option_case.rs
fn maybe_size(_key: Atom) -> NifResult<Option<f32>> {
    Ok((atoms::some(), 12.0))
}

fn apply_size(key: Atom) -> NifResult<f32> {
    let value = 1.0;
    if let Some(size) = maybe_size(key)? {
        let value = size;
    }
    Ok(value)
}

fn default_size(key: Atom) -> NifResult<f32> {
    Ok(maybe_size(key)?.unwrap_or(1.0))
}

fn guarded_result_case(key: Atom) -> NifResult<f32> {
    match maybe_size(key) {
        Ok(Some(size)) if size > 0.0 => Ok(size),
        _ => Ok(0.0),
    }
}
